### PR TITLE
Don’t error for missing await when promise is referenced in condition body

### LINKED ISSUE #43514: Inconsistent missing await in conditional errors, invalid quick fix
### 🔎 Search Terms

 * missing await in conditional
 * await
 * conditional

### 🕗 Version & Regression Information
Since 4.3.0-dev

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210403#code/CYUwxgNghgTiAEYD2A7AzgF0VMALEAXPAErhIzAA8mMAligOYA08ACjEgLa1ojUZ1GAPiEBYAFATQkWAmTosAMxAY8AFVz0GRABS1gRGloCU8ALxC2Hbr36CGYyePmZ4DFQGEc+YBq0BGc3g9A3gjRlMLeABvCXh4CBV4DE1GILBvEABtfQBdCTj4OAwAVxgUeABCFK14AHo6+BQkLCgKkBgOGBZaRWT8eA6u+Ho3T0zfVIYAZhG0eAA3KAh9fpAKgHdcAE9+njmUAHJW9s7yAH5C+PPgjLxsvKDlVVw-RhDjYyv4IhrGCQAvgVnKhXO4MF57pMtAAmIIhQwCEzmSyxcTxRJYP4MdKZHLAfJOeLFMoVbHwc43HR3fD43JPFTqKYfUwNJoteBtQZnGCA4EuLDgyE+N4zeH6RH2SKowqYvZpMzYe504HElSk+U4tlDcjBGEADgADP5jCwAEYlLC0LBoXBIEoQYDHeBmkAsIUTUWzHhHVpobYoMDwRQlQMYWiob43bHfXQ0h4EhkvUUsvniIA)

### 💻 Code

```ts
declare const cache: Record<string, Promise<string>>

declare const fetchThing: (id: string) => Promise<string>

const getCachedThing1 = (id: string) => {
  let thing = cache[id]

  return !thing // not an error, if the error in getCachedThing3 is valid then why this isn't an error?
    ? (cache[id] = fetchThing(id))
    : thing
}

const getCachedThing2 = (id: string) => {
  let thing = cache[id]

  return thing ?? (cache[id] = fetchThing(id)) // not an error
}

const getCachedThing3 = (id: string) => {
  let thing = cache[id]

  return thing // error (2801), but it should't be, getCachedThing3 isn't async function
    ? thing
    : (cache[id] = fetchThing(id))
}

```

After applying quick fix to `getChachedThing3` error:

```ts
const getCachedThing3 = (id: string) => {
  let thing = cache[id]

  return await thing // error (2801), but it should't be, getCachedThing3 isn't async function
 // error (2801), but it should't be, getCachedThing3 isn't async function
    ? thing
    : (cache[id] = fetchThing(id))
}
```

Quick fixed code is invalid (`await` is added to non-async function). Also, comment gets duplicated.

### 🙁 Actual behavior

1) Missing await in conditional errors are inconsistent
2) Missing await in conditional errors get reported in non-async functions
3) Quick fixing the error in non-async function produces invalid results

### 🙂 Expected behavior

1) Missing await in conditional errors to be consistent
2) Missing await in conditional errors to *NOT* be reported in non-async functions
3) No quick fix suggestion in non-async function

*** 

As a sidenote, I think the "missing await in conditional error" feature should be under a flag
1) It is wildly inconsistent
2) This is a job for linters
3) Objects are thruthy, so let them be thruthy :) 
4) This feature breaks existing code

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #43514

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -35392,12 +35383,13 @@ namespace ts {
             }
 
             // While it technically should be invalid for any known-truthy value
-            // to be tested, we de-scope to functions unrefenced in the block as a
-            // heuristic to identify the most common bugs. There are too many
-            // false positives for values sourced from type definitions without
-            // strictNullChecks otherwise.
+            // to be tested, we de-scope to functions and Promises unrefenced in
```
Comment (sandersn) [inline src/compiler/checker.ts:35386]:
```suggestion
            // to be tested, we de-scope to functions and Promises unreferenced in

```

**Review by sandersn — APPROVED** [review]
Maybe now is a good time to reword the "always returns ... since it appears to" errors?

**Comment by andrewbranch** [comment]
@DanielRosenwasser are you alright with the error rewording?

**Comment by DanielRosenwasser** [comment]
@andrewbranch I'm a little hurt that you want to undo all my hard work, but it's probably okay.

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by ejose19** [comment]
I think this broke original implementation, see [playground](https://www.typescriptlang.org/play?ts=4.3.2#code/IYZwngdgxgBAZgV2gFwJYHsL3egFASgC4YAFAJ3QFtUQBTAHgCMcAbW4CAPhgG8BYAFAxhMMrWQIyWOMBZ0A3IIC+gwaEixEKDFkbAyBYuSo0GIZGVQQA5t35CRqODFxwcBfL0EiYAel-CAH7Bwd4i-jAAomQUZACEYcIRACoAFjQwUJgAJqhomDAA7qgsLDCyhcBgIKLiklgWCLQwIFZQtIl+AcjpNQDkxtR0TKzsXH3lAA6T7GQ1yOjlLJXVMIzN2bRwVrTZAHSdEQAiqNkwYOgI2GTW4jALMAh0MH3AlXl9APydPmISUjAAESNWiAxQOYQqCG1f5YQEyOSg8FQwRAA) (as in [announcement](https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#truthy-promise-checks)). That exact code correctly errors out in beta [playground](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210403#code/IYZwngdgxgBAZgV2gFwJYHsL3egFASgC4YAFAJ3QFtUQBTAHgCMcAbW4CAPhgG8BYAFAxhMMrWQIyWOMBZ0A3IIC+gwaEixEKDFkbAyBYuSo0GIZGVQQA5t35CRqODFxwcBfL0EiYAel-CAH7Bwd4i-jAAomQUZACEYcIRACoAFjQwUJgAJqhomDAA7qgsLDCyhcBgIKLiklgWCLQwIFZQtIl+AcjpNQDkxtR0TKzsXH3lAA6T7GQ1yOjlLJXVMIzN2bRwVrTZAHSdEQAiqNkwYOgI2GTW4jALMAh0MH3AlXl9APydPmISUjAAESNWiAxQOYQqCG1f5YQEyOSg8FQoA).

**Comment by andrewbranch** [comment]
Well, that’s not great 😱

**Pull request merged**